source/lac/sparsity_tools.cc

```diff
@@ -374,85 +393,85 @@ namespace SparsityTools
-              std::vector<types::global_dof_index> &new_entries = groups.back();
-              new_entries.push_back(min_neighbors.first);
-              touched_nodes[min_neighbors.first] = groups.size()-1;
+              std::vector<types::global_dof_index> &next_group = groups.back();
 
-              // Add all direct neighbors of the pivot cell not yet touched to
-              // the current list
+              next_group.push_back(min_neighbors.first);
+              touched_nodes[min_neighbors.first] = groups.size()-1;
               for (DynamicSparsityPattern::row_iterator it
                    = connectivity.row_begin(min_neighbors.first);
                    it != connectivity.row_end(min_neighbors.first); ++it)
-                {
-                  if (touched_nodes[*it] == numbers::invalid_dof_index)
-                    {
-                      new_entries.push_back(*it);
-                      touched_nodes[*it] = groups.size()-1;
-                    }
-                }
+                if (touched_nodes[*it] == numbers::invalid_dof_index)
+                  {
+                    next_group.push_back(*it);
+                    touched_nodes[*it] = groups.size()-1;
+                  }
 
               // Add all neighbors of the current list not yet touched to the
-              // set of possible next pivots. Delete the entries of the current
-              // list from the set of possible next pivots.
-              for (types::global_dof_index i=0; i<new_entries.size(); ++i)
+              // set of possible next pivots. The added node is no longer a
+              // valid neighbor (here we assume symmetry of the
+              // connectivity). Delete the entries of the current list from
+              // the set of possible next pivots.
+              for (unsigned int i=0; i<next_group.size(); ++i)
                 {
                   for (DynamicSparsityPattern::row_iterator it
-                       = connectivity.row_begin(new_entries[i]);
-                       it != connectivity.row_end(new_entries[i]); ++it)
-                    if (touched_nodes[*it] == numbers::invalid_dof_index)
-                      current_neighbors.insert(*it);
-                  current_neighbors.erase(new_entries[i]);
+                       = connectivity.row_begin(next_group[i]);
+                       it != connectivity.row_end(next_group[i]); ++it)
+                    {
+                      if (touched_nodes[*it] == numbers::invalid_dof_index)
+                        current_neighbors.insert(*it);
+                      n_remaining_neighbors[*it]--;
+                    }
+                  current_neighbors.erase(next_group[i]);
                 }
             }
         }
-      // If the number of groups is smaller than the number of nodes, we can
+
+      // Sanity check: for all nodes, there should not be any neighbors left
+      for (types::global_dof_index row=0; row<connectivity.n_rows(); ++row)
```

Comment:
I think it probably should if `n_rows()` is `const`, as the body of the loop is empty in that case. I think we have lots of places like this in the library, but that doesn't preclude the idea that we should change to a different style if you think that's useful.